apps/theming/lib/Jobs/MigrateBackgroundImages.php

```diff
@@ -36,47 +35,82 @@
 use OCP\Files\NotPermittedException;
 use OCP\Files\SimpleFS\ISimpleFolder;
 use OCP\IConfig;
+use OCP\IDBConnection;
 
 class MigrateBackgroundImages extends QueuedJob {
 	public const TIME_SENSITIVE = 0;
 
+	protected const STAGE_PREPARE = 'prepare';
+	protected const STAGE_EXECUTE = 'execute';
+
 	private IConfig $config;
-	private IAppManager $appManager;
 	private IAppDataFactory $appDataFactory;
 	private IJobList $jobList;
-
-	public function __construct(ITimeFactory $time, IAppDataFactory $appDataFactory, IConfig $config, IAppManager $appManager, IJobList $jobList) {
+	private IDBConnection $dbc;
+
+	public function __construct(
+		ITimeFactory $time,
+		IAppDataFactory $appDataFactory,
+		IConfig $config,
+		IJobList $jobList,
+		IDBConnection $dbc
+	) {
 		parent::__construct($time);
 		$this->config = $config;
-		$this->appManager = $appManager;
 		$this->appDataFactory = $appDataFactory;
 		$this->jobList = $jobList;
+		$this->dbc = $dbc;
 	}
 
 	protected function run($argument): void {
-		if (!$this->appManager->isEnabledForUser('dashboard')) {
-			return;
+		if (!isset($argument['stage'])) {
+			// not executed in 25.0.0?!
+			$argument['stage'] = 'prepare';
 		}
 
-		$dashboardData = $this->appDataFactory->get('dashboard');
+		switch ($argument['stage']) {
+			case self::STAGE_PREPARE:
+				$this->runPreparation();
+				break;
+			case self::STAGE_EXECUTE:
+				$this->runMigration();
+				break;
+			default:
+				break;
+		}
+	}
 
-		$userIds = $this->config->getUsersForUserValue('theming', 'background', 'custom');
+	protected function runPreparation(): void {
+		try {
+			$selector = $this->dbc->getQueryBuilder();
+			$result = $selector->select('userid')
+				->from('preferences')
+				->where($selector->expr()->eq('appid', $selector->createNamedParameter('theming')))
+				->andWhere($selector->expr()->eq('configkey', $selector->createNamedParameter('background')))
+				->andWhere($selector->expr()->eq('configvalue', $selector->createNamedParameter('custom')))
+				->executeQuery();
+
+			$userIds = $result->fetchAll(\PDO::FETCH_COLUMN);
+			$this->storeUserIdsToProcess($userIds);
+		} catch (\Throwable $t) {
+			$this->jobList->add(self::class, self::STAGE_PREPARE);
+			throw $t;
+		}
+		$this->jobList->add(self::class, self::STAGE_EXECUTE);
+	}
 
-		$notSoFastMode = \count($userIds) > 5000;
-		$reTrigger = false;
-		$processed = 0;
+	protected function runMigration(): void {
+		$storedUserIds = $this->config->getAppValue('theming', '25_background_image_migration');
+		if ($storedUserIds === '') {
+			return;
+		}
+		$allUserIds = \json_decode(\gzuncompress($storedUserIds), true);
+		$notSoFastMode = isset($allUserIds[5000]);
```

Comment:
```suggestion
		$notSoFastMode = \count($allUserIds) > 5000;
```
Clearer what is being checked unless `isset` is faster then a comment could be added

Reply:
For one operation, I don't think speed really matter, I would use `count` event if `isset` take x100 more time.

Reply:
On 8.1, there is no significant difference, on PHP 7.4 isset is faster:

```
╚> php7 morethan.php 
time count(): 0.016927719116211ms
time isset(): 0.00095367431640625ms
17:58:07 work@zara:/dev/shm$
╚> php7 morethan.php 
time count(): 0.021934509277344ms
time isset(): 0.0019073486328125ms
17:58:12 work@zara:/dev/shm$
╚> php7 morethan.php 
time count(): 0.021934509277344ms
time isset(): 0.0021457672119141m
```

Reply:
https://3v4l.org/snZhB